Sources/MongoSwift/MongoCollection+Write.swift

```diff
@@ -35,46 +35,53 @@ extension MongoCollection {
     }
 
     /**
-     * Encodes the provided values to BSON and inserts them. If any values are missing identifiers,
-     * the driver will generate them.
+     * Encodes the provided values to BSON and inserts them. If any values are
+     * missing identifiers, the driver will generate them.
      *
      * - Parameters:
-     *   - documents: The `CollectionType` values to insert
-     *   - options: Optional `InsertManyOptions` to use when executing the command
+     *   - values: The `CollectionType` values to insert
+     *   - options: optional `InsertManyOptions` to use while executing the operation
      *
-     * - Returns: The optional result of attempting to performing the insert. If the write concern
-     *            is unacknowledged, nil is returned
+     * - Returns: an `InsertManyResult`, or `nil` if the write concern is unacknowledged.
+     *
+     * - Throws:
+     *   - `MongoError.invalidArgument` if `values` is empty
+     *   - `MongoError.insertManyError` if any error occurs while performing the writes
      */
     @discardableResult
     public func insertMany(_ values: [CollectionType], options: InsertManyOptions? = nil) throws -> InsertManyResult? {
-        let encoder = BsonEncoder()
+        if values.isEmpty {
+            throw MongoError.invalidArgument(message: "values cannot be empty")
+        }
 
+        let encoder = BsonEncoder()
+        let opts = try encoder.encode(options)
         let documents = try values.map { try encoder.encode($0) }
-        for doc in documents where !doc.keys.contains("_id") {
-            try ObjectId().encode(to: doc.storage, forKey: "_id")
+        var insertedIds: [Int: BsonValue?] = [:]
+
+        try documents.enumerated().forEach { (index, document) in
+            if !document.keys.contains("_id") {
```

Comment:
now that [`Document.hasKey`](https://github.com/mongodb/mongo-swift-driver/blob/185829de1358094b25838781c126db000cad8a7c/Sources/MongoSwift/BSON/Document.swift#L37) exists, we should use that in these cases to avoid reading all the keys out into Swift.